### ISSUE TITLE
Feature request: Support optional custom exception handling in Table.iter() + Table.read()

### ISSUE THREAD

**Issue by hjoukl** [frictionlessdata/tableschema-py#258]
# Overview

Feature request: Support custom exception handling through optional Table.iter() / Table.read() argument.

tableschema is great!
And tableschema-py can be easily used to great avail for data validation if it could be run in "fail-late" mode, i.e. not immediately raising exceptions in iter() or read(). As a generator function can't really be resumed if it raises a mechanism to gather errors instead of raising would be needed.

I therefore suggest adding an exc_handler option to these methods to support this:
```python
    def iter(self, keyed=False, extended=False, cast=True,                   
             integrity=False, relations=False,
             foreign_keys_values=False, exc_handler=None):
```
Current behaviour will remain unchanged if this `exc_handler` option is not used.

I started experimenting with this when I hadn't even realized there's also goodtables, which then had no FK-support.
While I realize that goodtables might be the intended go-to library for data validation I still think this would be a valuable addition:
- it can easily empower custom validation implementations on top of tableschema
- it's way faster in my experiments ;-) (like in 130s vs 350s for 3 csv files with 29257, 1514275,  488158 lines)

I'll submit a pull request for the sake of discussion as I think this is much easier than explaining it here in so many words.

Be warned: Should this be accepted in one form or another there'll be an follow-up PR to datapackage-py to make this opion usable throught the Package API ;-)

---

Please preserve this line to notify @roll (lead of this repository)

**Comment by roll** [comment]
Hi @hjoukl,

It sounds good! 

BTW have you tried https://github.com/frictionlessdata/goodtables-py, it can be useful as a validator too.

**Comment by hjoukl** [comment]
Hi @roll,

thank you for getting back!

> BTW have you tried https://github.com/frictionlessdata/goodtables-py, it can be useful as a validator too.

Yes, I've done that. It works just fine, but
- I'd like to have a bit more control over validating individual resources in a datapackage
- I'd like to be able to indicate validation progress. I'm doing this by instantiating a datapackage and then iterating over the separate resource with showing a nice tqdm (https://pypi.org/project/tqdm/) progress bar. To do so you need to receive feedback for each validated row, which is basically what this feature would allow.
- goodtables-py is currently way slower in my experiments: validating with datapackage + tableschema clocks in at 130s vs goodtables at 350s for 3 csv files with 29257, 1514275, 488158 lines

That said, I must admit I haven't yet looked very closely at goodtables-py because tableschema worked so fine for me with these little tweaks.  So maybe goodtables-py does indeed allow for picking individual resources from a datapackage for validation?

As to the performance I haven't measured where the difference actually comes from. I suspect the foreign key checking as one factor since this has become so fast in tableschema since #254 has been merged and released. My test data is rather FK-reference-heavy. Also, goodtables-py does produce a nicely streamlined json report which my validator script does not currently do to the same extent.

**Comment by roll** [comment]
@hjoukl 
You're right, `goodtables` needs optimization. I have written it down here - https://github.com/frictionlessdata/software/issues/38

**Comment by hjoukl** [comment]
Implemented with the merge of PR #259 in a87951c :-).
